File: ios/Freepass/Freepass/FieldViewModel.swift
import Foundation
import RxSwift

class FieldViewModel {
	enum FieldType {
		case Stored
		case Derived
	}

	let dbag = DisposeBag()
	let field_type: Variable<FieldType?> = Variable(nil)
	let field_name = Variable("")
	let derived_counter: Variable<UInt32> = Variable(0)
	let derived_site_name = Variable("")
	let derived_usage: Variable<DerivedUsage?> = Variable(nil)
	let stored_data: Variable<[UInt8]?> = Variable(nil)
	let stored_data_string = Variable("")
	let stored_usage: Variable<StoredUsage?> = Variable(nil)

	func init_stored_data_conversion() {
		var updatingFromSelf = false
		stored_data.asObservable().subscribeNext {
			if updatingFromSelf { return }
			let b = $0 ?? []
			if let s = NSString(bytes: b, length: b.count, encoding: NSUTF8StringEncoding) as? String {
				self.stored_data_string.value = s
			}
		}.addDisposableTo(dbag)
		stored_data_string.asObservable().subscribeNext {
			updatingFromSelf = true
			self.stored_data.value = $0.utf8.map { $0 }
			updatingFromSelf = false
		}.addDisposableTo(dbag)
	}

	init(name: String, field: Field) {
		init_stored_data_conversion()
		field_name.value = name
		switch field {
		case .Derived(let counter, let site_name, let usage):
			field_type.value = .Derived
			derived_counter.value = counter
			derived_site_name.value = site_name ?? ""
			derived_usage.value = usage
		case .Stored(data: let data, usage: let usage):
			field_type.value = .Stored
			stored_data.value = data
			stored_usage.value = usage
		}
	}

	func toField() -> (String, Field)? {
		switch (field_name.value, field_type.value) {
		case (let name, .Derived?):
			return (name, .Derived(counter: derived_counter.value ?? 1, site_name: derived_site_name.value, usage: derived_usage.value ?? .Password(template: "Maximum")))
		case (let name, .Stored?):
			return (name, .Stored(data: stored_data.value ?? [], usage: stored_usage.value ?? .Password))
		default: return nil
		}
	}
}

File: android/app/src/main/kotlin/technology/unrelenting/freepass/FieldViewModel.kt
package technology.unrelenting.freepass

import jp.keita.kagurazaka.rxproperty.RxProperty
import java.nio.charset.Charset

class FieldViewModel(name: String, field: Field?) {
	companion object {
		val UTF8 = Charset.forName("UTF-8")
	}

	enum class FieldType {
		Stored, Derived
	}

	val field_name = RxProperty(name)
	val field_type = RxProperty<FieldType?>()
	val field_type_radio = RxProperty(field_type.map {
		when (it) {
			FieldType.Stored -> R.id.fld_type_stored
			else -> R.id.fld_type_derived
		}
	}).let {
		it.subscribe {
			when (it) {
				R.id.fld_type_derived -> field_type.set(FieldType.Derived)
				R.id.fld_type_stored -> field_type.set(FieldType.Stored)
			}
		}
		it
	}
	val derived_counter = RxProperty<Int>(0)
	val derived_site_name = RxProperty<String>()
	val derived_usage = RxProperty<DerivedUsage?>()
	val stored_data = RxProperty<ByteArray?>()
	val stored_data_string = RxProperty<String>(stored_data.map {
		if (it == null) ""
		else try {
			String(it, UTF8)
		} catch (e: Exception) {
			""
		}
	}).let {
		it.subscribe {
			if (it != null && it.isNotEmpty())
				stored_data.set((it as java.lang.String).getBytes("UTF-8"))
		}
		it
	}
	val stored_usage = RxProperty<StoredUsage?>()

	init {
		if (field is Field.Derived) {
			field_type.set(FieldType.Derived)
			derived_counter.set(field.counter)
			derived_site_name.set(field.site_name.orEmpty())
			derived_usage.set(field.usage)
		} else if (field is Field.Stored) {
			field_type.set(FieldType.Stored)
			stored_data.set(field.data)
			stored_usage.set(field.usage)
		}
	}

	constructor(name: String) : this(name, null)
}
